File: LLD/oops.java
/* ABSTRACTION: using interface(implements), abstract-class(extends) */
abstract class Parent {
    void method() {
        System.out.println("Parent class method");
    }
}
interface oops {
    void method2();
}

class Child extends Parent implements oops{  /* INHERITANCE: single, hiarachial, multilevel  */
    void method() {
        System.out.println("Child class method");
    }

    @Override
    public void method2() { // polymorphism
    }
}

class Main {
    public static void main(String[] args) {
        Parent o2 = new Child(); // ENCAPSULATION
        o2.method();
    }
}

File: LLD/001oops/oops.cpp
#include <bits/stdc++.h>
using namespace std;

/* ABSTRACTION (abstract classes, interfaces)*/
class Parent {
    int data;
    public:
        Parent(int d) : data(d) {}
        virtual void func() = 0;
};

class Child : public Parent { /* INHERITANCE (single, multilevel, hierarchil(single parent), multiple(single child many parents)) */
public:
    Child(): Parent(0) {}

    void func() override{ /* POLYMORPHISM (runtime(method-overriding), compiletime(method-overloadin)) */
        cout << "Child class" << endl;
    }
};




int main() {
    Parent* ch = new Child(); /* ENCAPSULATION */
    ch->func();
}

